tests/cmsis_rtos_v2/src/thread_apis.c

```diff
@@ -247,3 +247,116 @@ void test_thread_prio(void)
 {
 	thread_prior_common(&thread3_state, &thread3_attr);
 }
+
+#define DELAY_MS 1000
+#define DELTA_MS 500
+
+static void thread5(void *argument)
+{
+	printk(" * Thread B started.\n");
+	osDelay(_ms_to_ticks(DELAY_MS));
+	printk(" * Thread B joining...\n");
+}
+
+static void thread4(void *argument)
+{
+	osThreadId_t tB = argument;
+	osStatus_t status;
+
+	printk(" + Thread A started.\n");
+	status = osThreadJoin(tB);
+	zassert_equal(status, osOK, "osThreadJoin thread B failed!");
+	printk(" + Thread A joining...\n");
+}
+
+void test_thread_join(void)
+{
+	osThreadAttr_t attr = { 0 };
+	s64_t time_stamp;
+	s64_t milliseconds_spent;
+	osThreadId_t tA, tB;
+	osStatus_t status;
+
+	attr.attr_bits = osThreadJoinable;
+
+	time_stamp = k_uptime_get();
+
+	printk(" - Creating thread B...\n");
+	tB = osThreadNew(thread5, NULL, &attr);
+	zassert_not_null(tB, "Failed to create thread B with osThreadNew!");
+
+	printk(" - Creating thread A...\n");
+	tA = osThreadNew(thread4, tB, &attr);
+	zassert_not_null(tA, "Failed to create thread A with osThreadNew!");
+
+	printk(" - Waiting for thread B to join...\n");
+	status = osThreadJoin(tB);
+	zassert_equal(status, osOK, "osThreadJoin thread B failed!");
+
+	if (status == osOK) {
+		printk(" - Thread B joined.\n");
+	}
+
+	milliseconds_spent = k_uptime_delta(&time_stamp);
+	zassert_true(milliseconds_spent >= DELAY_MS - DELTA_MS,
+		     "Join took too little time");
+	zassert_true(milliseconds_spent <= DELAY_MS + DELTA_MS,
+		     "Join took too much time.");
+
+	printk(" - Waiting for thread A to join...\n");
+	status = osThreadJoin(tA);
+	zassert_equal(status, osOK, "osThreadJoin thread A failed!");
+
+	if (status == osOK) {
+		printk(" - Thread A joined.\n");
+	}
+}
+
+void test_thread_detached(void)
+{
+	osThreadId_t thread;
+	osStatus_t status;
+
+	thread = osThreadNew(thread5, NULL, NULL); /* osThreadDetached */
+	zassert_not_null(thread, "Failed to create thread with osThreadNew!");
+
+	osDelay(_ms_to_ticks(DELAY_MS - DELTA_MS));
+
+	status = osThreadJoin(thread);
```

Comment:
In `test_thread_detached` `osThreadJoin` returns `osErrorResource` because the thread is no longer joinable.

In 'test_thread_terminate` `osThreadJoin` returns `osErrorResource` because the thread is no longer active.

From my interpretation of the CMSIS RTOS v2 API specification those would seem appropriate in those situations.

What do you think?

Reply:
Sounds good.